providers/implementations/digests/external_mu_ml_dsa_prov.c

```diff
@@ -0,0 +1,346 @@
+        return 0;
+    ctx->remaining = ctx->digest_len;
+    return mu_set_ctx_params(vctx, params);
+}
+
+static int mu_get_params(OSSL_PARAM params[])
+{
+    return ossl_digest_default_get_params(params, SHA3_BLOCKSIZE(256),
+                                          SHAKE256_SIZE, SHAKE_FLAGS);
+}
+
+static const OSSL_PARAM *mu_settable_ctx_params(ossl_unused void *ctx,
+                                                ossl_unused void *provctx)
+{
+    return external_mu_ml_dsa_set_ctx_params_list;
+}
+
+static int set_property_query(MU_CTX *ctx, const char *propq)
+{
+    OPENSSL_free(ctx->propq);
+    ctx->propq = NULL;
+    if (propq != NULL) {
+        ctx->propq = OPENSSL_strdup(propq);
+        if (ctx->propq == NULL)
+            return 0;
+    }
+    return 1;
+}
+
+static EVP_MD *shake_digest(MU_CTX *ctx)
+{
+    if (ctx->md == NULL)
+        ctx->md = EVP_MD_fetch(ctx->libctx, "SHAKE256", ctx->propq);
+    return ctx->md;
+}
+
+static int digest_public_key(MU_CTX *ctx, const uint8_t *pub, size_t publen)
+{
+    int ret;
+    EVP_MD *md;
+    EVP_MD_CTX *mdctx;
+
+    if (publen != ML_DSA_44_PUB_LEN
+            && publen != ML_DSA_65_PUB_LEN
+            && publen != ML_DSA_87_PUB_LEN) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
+        return 0;
+    }
+
+    md = shake_digest(ctx);
+    if (md == NULL)
+        return 0;
+    mdctx = EVP_MD_CTX_new();
+    if (mdctx == NULL)
+        return 0;
+    ret = EVP_DigestInit_ex(mdctx, md, NULL)
+        && EVP_DigestUpdate(mdctx, pub, publen)
+        && EVP_DigestFinalXOF(mdctx, ctx->tr, sizeof(ctx->tr));
+    EVP_MD_CTX_free(mdctx);
+
+    return ret;
+}
+
+static int mu_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    MU_CTX *ctx = (MU_CTX *)vctx;
+    struct external_mu_ml_dsa_set_ctx_params_st p;
+
+    if (ctx == NULL || !external_mu_ml_dsa_set_ctx_params_decoder(params, &p))
+        return 0;
+
+    if (p.ctx != NULL) {
+        void *vp = ctx->context;
+
+        if (!OSSL_PARAM_get_octet_string(p.ctx, &vp, sizeof(ctx->context),
+                                         &(ctx->context_len))) {
+            ctx->context_len = 0;
+            return 0;
+        }
+    }
+    if (p.propq != NULL) {
+        if (p.propq->data_type != OSSL_PARAM_UTF8_STRING
+            || !set_property_query(ctx, p.propq->data))
+            return 0;
+    }
+    if (p.pubkey != NULL) {
+        if (p.pubkey->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+        if (!digest_public_key(ctx, p.pubkey->data, p.pubkey->data_size))
+            return 0;
+        ctx->keylen = p.pubkey->data_size;
+    }
+    if (p.digestname != NULL) {
+        int ret;
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, p.digestname->data, ctx->propq);
+
+        if (md == NULL)
+            return 0;
+        ret = ossl_der_oid_pq_dsa_prehash_digest(md, &ctx->oid, &ctx->oid_len,
+                                                 &ctx->digest_len);
```

Comment:
This is doing a lot of work to fetch, inspect and discard an EVP_MD algorithm handle, when it could just check for the supported string names, and index into a fixed array of known DER OID blobs.
The strings in question are uniquely identified by the characters at offsets [3] and [6] which could switch into a case branch that checks the full string and sets the associated metadata.

None of this requires actually fetching the algorithm, since it is not actually used...